### DISCUSSION TITLE
Discuss wearable actuators implementation

### DISCUSSION THREAD

**Discussion by yeshasvitirupachuri** [robotology/wearables#103] [Ideas]
Currently the wearable library provides a means of encapsulation of sensor data through [various sensor interfaces](https://github.com/robotology/wearables/tree/master/interfaces/IWear/include/Wearable/IWear/Sensors) that can be packaged nicely into a wearable data format. This served us well in developing `wearable devices` such as [xsens suit](https://github.com/robotology/wearables/tree/master/devices/XsensSuit).

As we recently started working on development of other `wearable devices` such as the adaptive [Paexo](https://github.com/robotology/wearables/tree/master/devices/Paexo) exoskeleton from Ottobock, we realized that we need to have actuation related infrastructure too, e.g. setting the motor position for the adaptive Paexo. Another example case if the haptic device actuation, such as the mode of actuation that indicates the duration and vibration pattern as indicated in https://github.com/robotology/wearables/issues/100#issuecomment-750375890.  

So, given these recent example applications we came across, it would be nice to add the actuators implementation to the wearable library, that provides the high level interface to the client devices that connect to a `wearable device` and sends the actuation command as needed. An example case for such a scenario is the [Paexo](https://github.com/robotology/wearables/tree/master/devices/Paexo)  wearable device. At the moment it provides the data from the Paexo system as a wearable library sensor interface implementation https://github.com/robotology/wearables/pull/81. This information will be consumed by client devices such as [HumanStateProvider](https://github.com/robotology/human-dynamics-estimation/tree/master/devices/HumanStateProvider), that in turn would like to send an actuation command to the Paexo actuator based on the adaptive control algorithms we implement.

**Comment by yeshasvitirupachuri** [comment]
I realize that a template for the implementation for handling actuation is already available in the wearables library. 

[IXsensMVNControl](https://github.com/robotology/wearables/blob/master/interfaces/IXsensMVNControl/IXsensMVNControl.h) interface provides the template for the utility functions that are implemented in [XsensSuit](https://github.com/robotology/wearables/blob/master/devices/XsensSuit/src/XsensSuit.cpp) wearable device, and [IXsensMVNControlWrapper](https://github.com/robotology/wearables/blob/master/wrappers/IXsensMVNControl/src/IXsensMVNControlWrapper.cpp) wrapper that opens a yarp rpc port to received calibration related commands. 

Along the same lines, one possible implementation for actuators is `IActuators` which is on top of actuators like `IHaptic`. A wrapper like `IActuatorsWrapper` opens a rpc port to receive inputs for the actuation. 

![photo_2021-01-11_13-22-23](https://user-images.githubusercontent.com/6505998/104182013-1268cc80-5410-11eb-998c-2b646a80c049.jpg)

**Comment by lrapetti** [comment]
A consideration here is, why those actuators should communicate via `rpc`? I guess the main reason is that we want a feedback from the actuators to make sure the command is received.

But what if the actuators should be set continuously, at high frequency? In this case what we already have for WearableData would work as well for sending the actuator command

**Comment by yeshasvitirupachuri** [comment]
> A consideration here is, why those actuators should communicate via rpc?

In some of the use cases we have now e,g, Paexo wearable device, the actuation commends will be generated probable by another process e.g. at the consumer layer of the architecture https://github.com/robotology/human-dynamics-estimation/blob/master/misc/hde_scheme.png.  Assuming the actuation can be asynchronous, unlike the sensor data streaming done through `IWear` interface and `IWearWrapper`, yarp rpc is the choice.

  > But what if the actuators should be set continuously, at high frequency? In this case what we already have for WearableData would work as well for sending the actuator command

Can you please explain this further. 

One important consideration we need to account for here is, any change in the update of wearable structure (by adding new actuators commands as part of wearableData structure) will render the wearableData datasets we have unusable, as we experienced when adding new sensors with iCub wearable device https://github.com/robotology/wearables/pull/25. So, we need to make the actuators part optional.

**Comment by yeshasvitirupachuri** [comment]
Update from meeting on 11th Jan 2020

Participants: @diegoferigo @lrapetti @evalli-iit @traversaro 

#### Action Points

- Decide on the interface methods for the actuators and implement them 
- Use streaming port over `rpc` for the actuators wrapper
- A remapper is not strictly necessary, and the client device (e.g. adaptive control algorithm application for Paexo) can connect to the actuators wrapper and send commands
- As thrift does not allow to optional data for [WearableData](https://github.com/robotology/wearables/blob/master/msgs/thrift/WearableData.thrift#L173), it is better to keep the actuators implementation separate from `IWear` interface
